### PR TITLE
Added Spring Boot 3 support

### PR DESCRIPTION
**Pull request by wllianwd**
# Changes
* Updated `Java` to `17`
* Updated `Spring Boot` to `3.0.1`
* Updated `Spring Cloud` to `2022.0.0`
* Updated to use `httpclient5` instead of `httpclient`
* Updated to use `log4j-slf4j2-impl` instead of `log4j-slf4j-impl`
* Removed sleuth as it was [replaced by micrometer tracing](https://spring.io/blog/2022/10/12/observability-with-spring-boot-3)

Can someone with write permissions take a look?

### CODE REVIEW

**Comment by jonathannaguin** [comment]
@kptfh will this PR be merged?

**Comment by alex-tilkin** [comment]
@wllianwd, joining @jonathannaguin question. Thank you for your effort

**Comment by ghost** [comment]
When this is planned to be merged please?

**Comment by CristianoPassos** [comment]
Same here, could some maintainer please have a look into it?

**Comment by wllianwd** [comment]
Trying to tag someone from other playtika projects.
@matveym

**Comment by wllianwd** [comment]
Trying to tag some from other Playtika projects: @Fameing

**Comment by Fameing** [comment]
I don't have access to this repository. I am trying to contact @kptfh. 
I will let you know how the result will be.

> Trying to tag some from other Playtika projects: @Fameing

File: feign-reactor-core/pom.xml
Comment (kptfh) [inline feign-reactor-core/pom.xml]: may try to set test scope?

File: feign-reactor-core/src/test/java/reactivefeign/resttemplate/client/RestTemplateFakeReactiveFeign.java
Comment (kptfh) [inline feign-reactor-core/src/test/java/reactivefeign/resttemplate/client/RestTemplateFakeReactiveFeign.java]: please format this file in separate PR
it's difficult to see what was changed

**Comment by matveym** [comment]
Hello @wllianwd. I do not think it will be possible to merge this PR as is - we need to maintain backward compatibility with Java 11/Spring Boot 2. I guess this means that we have to release two versions of feign-reactive.
We are working on the solution, sorry for delay.

**Comment by CristianoPassos** [comment]
> Hello @wllianwd. I do not think it will be possible to merge this PR as is - we need to maintain backward compatibility with Java 11/Spring Boot 2. I guess this means that we have to release two versions of feign-reactive. We are working on the solution, sorry for delay.

Let me know if you guys need any help, I would like to contribute.

**Comment by wllianwd** [comment]
> Hello @wllianwd. I do not think it will be possible to merge this PR as is - we need to maintain backward compatibility with Java 11/Spring Boot 2. I guess this means that we have to release two versions of feign-reactive. We are working on the solution, sorry for delay.

Hey @matveym, I updated the modules to version `4.0.0` as my guess was:
* `4.x.x` would be Java `17`, Spring Boot `3`, using main branch
* `3.x.x` would be Java `11`, Spring Boot `2`, can use a `3.x` branch or be created out the latest `3.x` tag, if needed

**Comment by matveym** [comment]
@wllianwd I think we should think through how to maintain the two versions in the future - backport changes or have some shared core. Maybe it would be better to create two separate Maven modules in the main branch.
It's up to the main maintainer - @kptfh to decide.
He's a bit busy now at work, will look into this later.
It is not urgent for you, isn't it?

**Comment by RoopeHakulinen** [comment]
Depends on the definition of urgent, I guess. In our case, it is the single dependency that has kept us from upgrading to Spring Boot 3 and Spring 6 for the past few months. As such, would be superb to get it merged, but I appreciate there are other priorities, too, and this is an open source library :)

**Comment by wllianwd** [comment]
> @wllianwd I think we should think through how to maintain the two versions in the future - backport changes or have some shared core. Maybe it would be better to create two separate Maven modules in the main branch.
It's up to the main maintainer - @kptfh to decide.
He's a bit busy now at work, will look into this later.
It is not urgent for you, isn't it?

As far I can see in several projects (Spring Cloud, Spring Cloud AWS, etc), the Spring Boot 3 changes will require you to migrate your code, I don't see that is possible/recommended to maintain Spring Boot 2 and 3 in the same branch. Also migration from Java 11 to 17 is required. 

Usually a different release train branch is created, like a `3.x.x`. In case you need to maintain the legacy version, you can use this branch.

For us it is also blocking the upgrade to Spring Boot 3.

**Comment by yangwudong** [comment]
> 
Spring released 3.x as beta almost half year ago, but until now, this lib still can't accept this PR to fit Spring Boot 3.x even Spring Boot 3 was official release several months.

Just my two cents, I think that's why we usually choose all Spring provided lib instead of community-offered cases.
However, I still like this feign-reactive, but if it can't accept/respond to the community frequently, I might choose WebClient instead of this, even it is quite ugly.

**Comment by stephweiss** [comment]
I would appreciate this feature :)

**Comment by dkulig** [comment]
Huge thumbs up for having this resolved

**Comment by matveym** [comment]
Hello all. Quick update - we have this work planned, but there would be an additional delay. As I was told we are moving now all our projects to GitHub Actions for build, have to finish this activity first.
Will keep you updated.
Sorry for the mess.

**Comment by DanishDeveloper360** [comment]
Could you please give us an ETA ?

**Comment by matveym** [comment]
Sorry, no ETA for now. It turned out we need to urgently migrate to GHA, here's the PR blocking this one: https://github.com/PlaytikaOSS/feign-reactive/pull/556.

**Comment by martypitt** [comment]
FYI - we needed this, as it's blocking us.

I took @wllianwd 's awesome PR, have forked it, and published a build - only as a short-term measure.
Once this PR gets merged, we'll kill it.

However, to avoid any conflicts around versioning, we've published our temp build under our own package, and in our own (public) maven repo.

```
      <repository>
         <id>orbital-snapshots</id>
         <url>https://repo.orbitalhq.com/snapshot</url>
         <snapshots>
            <enabled>true</enabled>
         </snapshots>
      </repository>
      <repository>
         <id>orbital-snapshots</id>
         <url>https://repo.orbitalhq.com/release</url>
      </repository>
```

And the dependency:

```
      <dependency>
         <groupId>com.orbitalhq.reactivefeign</groupId>
         <artifactId>feign-reactor-webclient</artifactId>
        <version>4.0.0-SNAPSHOT</version>
      </dependency>
```

We'll publish a release version shortly.

Code is [here](https://github.com/orbitalapi/feign-reactive)

**Comment by DanishDeveloper360** [comment]
> FYI - we needed this, as it's blocking us.
> 
> I took @wllianwd 's awesome PR, have forked it, and published a build - only as a short-term measure.
> Once this PR gets merged, we'll kill it.
> 
> However, to avoid any conflicts around versioning, we've published our temp build under our own package, and in our own (public) maven repo.
> 
> ```
>       <repository>
>          <id>orbital-snapshots</id>
>          <url>https://repo.orbitalhq.com/snapshot</url>
>          <snapshots>
>             <enabled>true</enabled>
>          </snapshots>
>       </repository>
>       <repository>
>          <id>orbital-snapshots</id>
>          <url>https://repo.orbitalhq.com/release</url>
>       </repository>
> ```
> 
> And the dependency:
> 
> ```
>       <dependency>
>          <groupId>com.orbitalhq.reactivefeign</groupId>
>          <artifactId>feign-reactor-webclient</artifactId>
>         <version>4.0.0-SNAPSHOT</version>
>       </dependency>
> ```
> 
> We'll publish a release version shortly.
> 
> Code is [here](https://github.com/orbitalapi/feign-reactive)

Hi there, are you/team committed for any bugs or CVEs in the short term public maven repo snapshot version? And what would you suggest about using this in enterprise companies which are blocked same as you are ?

**Comment by DanishDeveloper360** [comment]
> Sorry, no ETA for now. It turned out we need to urgently migrate to GHA, here's the PR blocking this one: https://github.com/PlaytikaOSS/feign-reactive/pull/556.

Thank you for you response, same as many other users of the library we are also blocked.

**Comment by martypitt** [comment]
> > FYI - we needed this, as it's blocking us.
> > I took @wllianwd 's awesome PR, have forked it, and published a build - only as a short-term measure.
> > Once this PR gets merged, we'll kill it.
> > However, to avoid any conflicts around versioning, we've published our temp build under our own package, and in our own (public) maven repo.
> > ```
> >       <repository>
> >          <id>orbital-snapshots</id>
> >          <url>https://repo.orbitalhq.com/snapshot</url>
> >          <snapshots>
> >             <enabled>true</enabled>
> >          </snapshots>
> >       </repository>
> >       <repository>
> >          <id>orbital-snapshots</id>
> >          <url>https://repo.orbitalhq.com/release</url>
> >       </repository>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And the dependency:
> > ```
> >       <dependency>
> >          <groupId>com.orbitalhq.reactivefeign</groupId>
> >          <artifactId>feign-reactor-webclient</artifactId>
> >         <version>4.0.0-SNAPSHOT</version>
> >       </dependency>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We'll publish a release version shortly.
> > Code is [here](https://github.com/orbitalapi/feign-reactive)
> 
> Hi there, are you/team committed for any bugs or CVEs in the short term public maven repo snapshot version? And what would you suggest about using this in enterprise companies which are blocked same as you are ?

No, not at this time.

You're welcome to raise PRs against it, and we'll probably merge them in, providing the code is good quality and passes tests.

But we're not looking to become long term contributors just yet.

That said, Reactive Feign and spring tooling are closely related to our core product, so we might revisit and explore contributing more formally later.

File: feign-reactor-parent/pom.xml
Comment (ijusti) [inline feign-reactor-parent/pom.xml:22]:
```suggestion
        <spring-cloud-bom.version>2022.0.1</spring-cloud-bom.version>
        <spring-boot-dependencies-bom.version>3.0.5</spring-boot-dependencies-bom.version>
```

File: feign-reactor-spring-configuration/src/main/resources/META-INF/spring.factories
Comment (ijusti) [inline feign-reactor-spring-configuration/src/main/resources/META-INF/spring.factories:1]: do not delete this file

**Review by ijusti — DISMISSED** [review]

File: feign-reactor-spring-configuration/src/main/resources/META-INF/spring.factories
Comment (wllianwd) [reply feign-reactor-spring-configuration/src/main/resources/META-INF/spring.factories:1]: This file is not supported anymore:
https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-3.0-Migration-Guide#auto-configuration-files

**Comment by DanishDeveloper360** [comment]
> FYI - we needed this, as it's blocking us.
> 
> I took @wllianwd 's awesome PR, have forked it, and published a build - only as a short-term measure. Once this PR gets merged, we'll kill it.
> 
> However, to avoid any conflicts around versioning, we've published our temp build under our own package, and in our own (public) maven repo.
> 
> ```
>       <repository>
>          <id>orbital-snapshots</id>
>          <url>https://repo.orbitalhq.com/snapshot</url>
>          <snapshots>
>             <enabled>true</enabled>
>          </snapshots>
>       </repository>
>       <repository>
>          <id>orbital-snapshots</id>
>          <url>https://repo.orbitalhq.com/release</url>
>       </repository>
> ```
> 
> And the dependency:
> 
> ```
>       <dependency>
>          <groupId>com.orbitalhq.reactivefeign</groupId>
>          <artifactId>feign-reactor-webclient</artifactId>
>         <version>4.0.0-SNAPSHOT</version>
>       </dependency>
> ```
> 
> We'll publish a release version shortly.
> 
> Code is [here](https://github.com/orbitalapi/feign-reactive)

Thank you, could you please let me know when is the plan for releasing this ?

**Review by Fameing — DISMISSED** [review]

**Comment by DanishDeveloper360** [comment]
@martypitt could you please let us know when will this be published to orbital ?

**Comment by martypitt** [comment]
It should be available in the snapshot repository now.

If you need a release version, can probably do tonight.

**Comment by DanishDeveloper360** [comment]
@martypitt me and my team would really appreciate if you'd provide us with a release version thank you.

**Comment by martypitt** [comment]
> @martypitt me and my team would really appreciate if you'd provide us with a release version thank you.

Sure - a 4.0.0 release is now available in:

```
      <repository>
         <id>orbital</id>
         <url>https://repo.orbitalhq.com/release</url>
      </repository>
```

With the maven co-ordinates of:

```
      <dependency>
         <groupId>com.orbitalhq.reactivefeign</groupId>
         <artifactId>feign-reactor-webclient</artifactId>
        <version>4.0.0</version>
      </dependency>
```

Just to reiterate - as soon as the official Playtika image is released, you should migrate across to that.  We're not really planning on support, we just needed to bridge the gap.

**Comment by MazizEsa** [comment]
Hi all anything I can help out with this? @kptfh  @wllianwd  ?

**Comment by matveym** [comment]
@kptfh is already working on that, wait a little, please. We need to resolve some issues first.

**Comment by MazizEsa** [comment]
@matveym dont mean to rush. I am sincere to help if there's anything needed :). All the best to everyone from Ukraine.

**Comment by DanishDeveloper360** [comment]
> Sorry, no ETA for now. It turned out we need to urgently migrate to GHA, here's the PR blocking this one: https://github.com/PlaytikaOSS/feign-reactive/pull/556.

@matveym Good news : 🎉, the blocking PR is merged https://github.com/PlaytikaOSS/feign-reactive/pull/556.

**Comment by ghost** [comment]
Can we move forward with this please? What is missing? I can help if needed.

**Comment by wllianwd** [comment]
@kptfh changes were done, [build](https://app.circleci.com/pipelines/github/PlaytikaOSS/feign-reactive/1815/workflows/6f5d8d6b-9718-46e1-92d7-223c77369508/jobs/1752) is passing, can we merge this?

**Comment by kptfh** [comment]
This branch cannot be rebased due to conflicts

**Comment by wllianwd** [comment]
@kptfh I resolved the conflicts, here it says `This branch has no conflicts with the base branch`.

Can you check again? I'm targeting `develop`.

**Review by kptfh — APPROVED** [review]

**Comment by SergeVolodin** [comment]
Good Day. We are migrating the project to spring boot 3.2 and encountered an error when the feign client (our feign interface which annotated by  @ReactiveFeignClient (reactivefeign.spring.config.ReactiveFeignClient)) is registered incorrectly in the FactoryBean. The error occurs during the attribute check step in org.springframework.beans.factory.support when executing ResolvableType getTypeForFactoryBeanFromAttributes(AttributeAccessor attributes) ;
As a result we get an IllegalArgumentException
"Invalid value type for attribute 'factoryBeanObjectType': java.lang.String".
Could you please help us to solve this problem?

Thanks.

**Comment by SergeVolodin** [comment]
found the discussion here 
https://github.com/PlaytikaOSS/feign-reactive/issues/663

**Pull request merged**
